primer-miso/src/Primer/Miso/Util.hs

```diff
@@ -250,3 +286,89 @@ nodeSelectionType =
           THole{} -> True
           TEmptyHole{} -> True
           _ -> False
+
+-- TODO keep an eye out for whether these ever return `Nothing` in practice
+-- the current hypothesis is that, because we already typecheck on initialisation and on performing each action,
+-- the app only ever deals with typechecked expressions
+-- and that it is therefore possible that we could refactor the library to assert this in the Haskell types
+-- note that we try to always work with typechecked expressions in the frontend,
+-- as being able to assume that metadata is always available makes the core application code simpler
+assumeDefHasTypeCheckInfo :: ASTDef -> Maybe ASTDefT
+assumeDefHasTypeCheckInfo def = do
+  expr <- sequenceOf (_exprMeta % _type) (astDefExpr def) >>= sequenceOf (_exprTypeMeta % _type)
+  sig <- sequenceOf (_typeMeta % _type) (astDefType def)
+  pure ASTDefT{expr, sig}
+assumeDefSelectionHasTypeCheckInfo :: DefSelection (Either ExprMeta (Either TypeMeta KindMeta)) -> Maybe DefSelectionT
+assumeDefSelectionHasTypeCheckInfo =
+  traverseOf #node $ traverse $ traverseOf #meta $ bitraverse (sequenceOf _type) $ bitraverse (sequenceOf _type) pure
+
+-- see `assumeDefHasTypeCheckInfo`
+-- sometimes we need to discard that extra type-level information,
+-- in order to get an input for various Primer library functions
+astDefTtoAstDef :: ASTDefT -> ASTDef
+astDefTtoAstDef def = ASTDef{astDefExpr = exprTtoExpr def.expr, astDefType = typeTtoType def.sig}
+
+-- this is potentially a better API then the one which `Primer.Action.Available` currently exports
```

Comment:
Note to self: look at `Primer.Action.Available` and see if we should refactor this.